torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,602 @@
+        } else if ((val(gtd_new) * (bracket[high_pos] - bracket[low_pos])) >= 0) {
+          // old high becomes new low
+          bracket[high_pos] = bracket[low_pos];
+          bracket_f[high_pos] = bracket_f[low_pos];
+          bracket_g[high_pos] = bracket_g[low_pos];
+          bracket_gtd[high_pos] = bracket_gtd[low_pos];
+        }
 
+        // new point becomes new low
+        bracket[low_pos] = t;
+        bracket_f[low_pos] = f_new;
+        bracket_g[low_pos] = g_new.clone(at::MemoryFormat::Contiguous);
+        bracket_gtd[low_pos] = gtd_new;
+      }
+
+      // line-search bracket is so small
+      if ((abs(bracket[1] - bracket[0]) * d_norm) < tolerance_change) break;
+    }
+
+    // return stuff
+    t = bracket[low_pos];
+    f_new = bracket_f[low_pos];
+    g_new = bracket_g[low_pos];
+    return std::make_tuple(f_new, g_new, t, ls_func_evals);
+}
+
+Tensor LBFGS::step(LossClosure closure) {
+  NoGradGuard no_grad;
+  TORCH_CHECK(closure != nullptr, "LBFGS requires a closure function");
+  TORCH_INTERNAL_ASSERT(param_groups_.size() == 1);
+  auto val = [](const Tensor& t) {return t.item<double>(); };
+
+  auto& group = param_groups_.at(0);
+  auto& _params = group.params();
+  const auto& options = static_cast<const LBFGSOptions&>(group.options());
+  auto lr = options.lr();
+  auto max_iter = options.max_iter();
+  auto max_eval = options.max_eval();
+  auto tolerance_grad = options.tolerance_grad();
+  auto tolerance_change = options.tolerance_change();
+  auto line_search_fn = options.line_search_fn();
+  auto history_size = options.history_size();
+
+  // NOTE: LBFGS has only global state, but we register it as state for
+  // the first param, because this helps with casting in load_state_dict
+  auto param_state = state_.find(c10::guts::to_string(_params.at(0).unsafeGetTensorImpl()));
+  if (param_state == state_.end()) {
+    state_[c10::guts::to_string(_params.at(0).unsafeGetTensorImpl())] = std::make_unique<LBFGSParamState>();
+  }
+  auto& state = static_cast<LBFGSParamState&>(*state_[c10::guts::to_string(_params.at(0).unsafeGetTensorImpl())]);
+  // evaluate initial f(x) and df/dx
+  Tensor orig_loss;
+  {
+    torch::AutoGradMode enable_grad(true);
+    orig_loss = closure();
+  }
+
+  auto loss = orig_loss.clone(at::MemoryFormat::Contiguous);
+  auto current_evals = 1;
+  state.func_evals(state.func_evals() + 1);
+  auto flat_grad = _gather_flat_grad();
+  auto opt_cond = (val(flat_grad.abs().max()) <= tolerance_grad);
+
+  // optimal condition
+  if (opt_cond) {
+    return orig_loss;
+  }
+
+  // tensors cached in state (for tracing)
+  auto& d = state.d();
+  auto& t = state.t();
+  auto& old_dirs = state.old_dirs();
+  auto& old_stps = state.old_stps();
+  auto& ro = state.ro();
+  auto& H_diag = state.H_diag();
+  auto& prev_flat_grad = state.prev_flat_grad();
+  auto& prev_loss = state.prev_loss();
+  auto& al = state.al();
+
+  int n_iter = 0;
   Tensor ONE = torch::tensor(1, flat_grad.options());
```

Comment:
This seems to be only used once, and I believe we can do `H_diag = torch::tensor(1)` at the use site instead.